### PR TITLE
[SPARK-32537][SQL][TEST] Add a CTEHintSuite for test coverage

### PR DESCRIPTION
**Pull request by LantaoJin**
### What changes were proposed in this pull request?
Add a new test suite `CTEHintSuite`

### Why are the changes needed?
This ticket is to address the below comments to help us understand the test coverage of SQL HINT for CTE.
https://github.com/apache/spark/pull/29062#discussion_r463247491
https://github.com/apache/spark/pull/29062#discussion_r463248167

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Add a test suite.

### CODE REVIEW

**Comment by LantaoJin** [comment]
cc @gatorsmile @cloud-fan

File: sql/core/src/test/scala/org/apache/spark/sql/CTEHintSuite.scala [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql
+
+import org.apache.log4j.Level
+
+import org.apache.spark.sql.catalyst.plans.PlanTest
+import org.apache.spark.sql.catalyst.plans.logical.{BROADCAST, HintInfo, Join, JoinHint, Repartition, RepartitionByExpression, ResolvedHint, SHUFFLE_HASH, SHUFFLE_MERGE, SHUFFLE_REPLICATE_NL}
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/CTEHintSuite.scala:23]: nit: I think its okay to fold it: `import org.apache.spark.sql.catalyst.plans.logical._`

File: sql/core/src/test/scala/org/apache/spark/sql/CTEHintSuite.scala [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+import org.apache.spark.sql.execution.adaptive.AdaptiveSparkPlanHelper
+import org.apache.spark.sql.test.SharedSparkSession
+
+class CTEHintSuite extends PlanTest with SharedSparkSession with AdaptiveSparkPlanHelper {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/CTEHintSuite.scala:27]: we need `AdaptiveSparkPlanHelper`?

File: sql/core/src/test/scala/org/apache/spark/sql/CTEHintSuite.scala [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+  def verifyCoalesceHint(df: DataFrame): Unit = {
+    val optimized = df.queryExecution.optimizedPlan
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/CTEHintSuite.scala:30]: Could you test it in an analyzed plan, too?

File: sql/core/src/test/scala/org/apache/spark/sql/CTEHintSuite.scala [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+    val repartitions = optimized collect {
+      case r: Repartition => r
+      case r: RepartitionByExpression => r
+      case _: ResolvedHint => fail("ResolvedHint should not appear after optimize.")
+    }
+    assert(repartitions.nonEmpty)
+  }
+
+  def verifyJoinHint(df: DataFrame, expectedHints: Seq[JoinHint]): Unit = {
+    val optimized = df.queryExecution.optimizedPlan
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/CTEHintSuite.scala:40]: ditto

File: sql/core/src/test/scala/org/apache/spark/sql/CTEHintSuite.scala (hunk 2 above)
Comment (LantaoJin) [reply sql/core/src/test/scala/org/apache/spark/sql/CTEHintSuite.scala:27]: I will remove it. I just copy it from `JoinHintSuite` :P

File: sql/core/src/test/scala/org/apache/spark/sql/CTEHintSuite.scala [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+            |)
+            |SELECT * FROM cte
+            |""".stripMargin),
+        JoinHint(
+          Some(HintInfo(strategy = Some(SHUFFLE_REPLICATE_NL))),
+          Some(HintInfo(strategy = Some(SHUFFLE_HASH)))) :: Nil
+      )
+    }
+  }
+}
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/CTEHintSuite.scala:158]: can we move https://github.com/apache/spark/pull/29062/files#diff-5a2e7f03d14856c8769fd3ddea8742bdR3564 to this new suite?

File: sql/core/src/test/scala/org/apache/spark/sql/CTEHintSuite.scala [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+package org.apache.spark.sql
+
+import org.apache.log4j.Level
+
+import org.apache.spark.sql.catalyst.plans.logical._
+import org.apache.spark.sql.test.SharedSparkSession
+
+class CTEHintSuite extends QueryTest with SharedSparkSession {
+
+  def verifyCoalesceHint(df: DataFrame): Unit = {
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/CTEHintSuite.scala:27]: nit: `Coalesce` -> `Repartition`?

File: sql/core/src/test/scala/org/apache/spark/sql/CTEHintSuite.scala [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+    def checkPlanContainsRepartition(plan: LogicalPlan): Unit = {
+      val repartitions = plan collect {
+        case r: Repartition => r
+        case r: RepartitionByExpression => r
+        case _: ResolvedHint => fail("ResolvedHint should not appear after optimize.")
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/CTEHintSuite.scala:32]: Is this error message correct even if the test is for an analyzed plan? https://github.com/apache/spark/pull/29359/files#diff-0d893e4fe9a621d3e084c2913c533e00R38

File: sql/core/src/test/scala/org/apache/spark/sql/CTEHintSuite.scala [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+  }
+
+  test("SPARK-32237: Hint in CTE") {
+    withTable("t") {
+      sql("CREATE TABLE t USING PARQUET AS SELECT 1 AS id")
+      checkAnswer(
+        sql(s"""
+               |WITH cte AS (SELECT /*+ REPARTITION(3) */ * FROM t)
+               |SELECT * FROM cte
+        """.stripMargin),
```
Comment (maropu) [inline sql/core/src/test/scala/org/apache/spark/sql/CTEHintSuite.scala:174]: nit format: 
```
      checkAnswer(
        sql(
          s"""
             |WITH cte AS (SELECT /*+ REPARTITION(3) */ * FROM t)
             |SELECT * FROM cte
             |""".stripMargin),
        Row(1) :: Nil)
```

File: sql/core/src/test/scala/org/apache/spark/sql/CTEHintSuite.scala (hunk 6 above)
Comment (LantaoJin) [reply sql/core/src/test/scala/org/apache/spark/sql/CTEHintSuite.scala:27]: `ResolveCoalesceHints` accepts names "COALESCE", "REPARTITION", and "REPARTITION_BY_RANGE".
https://github.com/apache/spark/blob/9a35b93c8a02939bf02437f47501d55223b6a283/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveHints.scala#L174

Comment (LantaoJin) [reply sql/core/src/test/scala/org/apache/spark/sql/CTEHintSuite.scala]: Changed to `def verifyCoalesceOrRepartitionHint`

**Comment by LantaoJin** [comment]
retest this please

**Review by maropu — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/CTEHintSuite.scala [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+            |SELECT * FROM cte
+            |""".stripMargin),
+        JoinHint(
+          Some(HintInfo(strategy = Some(SHUFFLE_REPLICATE_NL))),
+          Some(HintInfo(strategy = Some(SHUFFLE_HASH)))) :: Nil
+      )
+    }
+  }
+
+  test("SPARK-32237: Hint in CTE") {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/CTEHintSuite.scala:169]: Actually we can remove it now. It's already covered by `Resolve coalesce hint in CTE`.

**Comment by cloud-fan** [comment]
github action passed, merging to master, thanks!

**Pull request closed without merging**